### PR TITLE
[Automodel] Fix CP device_mesh issue, use PTL distsampler

### PR DESCRIPTION
**Pull request by BoxiangW**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 1]
```diff
@@ -371,6 +372,17 @@ def training_step(self, batch, batch_idx=None, context_parallel=False):
                 labels = labels.view(-1)
                 assert logits.shape[-2] == labels.shape[-1], "Expected logits & labels to have the same length"
                 loss = self.loss_fn(logits, labels, loss_mask)
+
+                num_tokens = loss_mask.sum().float()
+                # change nan to 0, this is for CP to ignore loss for padded tokens
+                loss = torch.nan_to_num(loss, nan=0.0)
```
Comment (cspades) [inline nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:378]: Shouldn't we just multiply the loss by the loss mask to zero out the padded token losses? Cause this will just change all NaN to 0.0 right?

Comment (cspades) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:378]: torch.where(bool(loss_mask), loss, zeros) was what I was thinking of!

Comment (cspades) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:378]: Also, can we move this down to the bottom so it applies to all 3 if-else branches? The logic is valid for all of them!

Comment (cspades) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:378]: Actually, the NaN just comes from dividing by zero when reducing with the mean.

We can just add this to the end:
```
# In the case where all labels are masked, the loss should be 0.
        if loss_mask is not None and loss_mask.bool().sum() == 0:
            loss.detach().copy_(torch.zeros_like(loss))
```

**Review by cspades — APPROVED** [review]

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 2]
```diff
@@ -436,11 +443,7 @@ def on_before_optimizer_step(self, optimizer) -> None:
                 # Use the flattened DP / CP device mesh for loss reduction
                 # if it exists (CP > 1), else default to the data parallel mesh.
                 group = device_mesh[
-                    (
-                        "dp_cp"
-                        if device_mesh.mesh_dim_names is not None and "dp_cp" in device_mesh.mesh_dim_names
-                        else "data_parallel"
-                    )
+                    ("dp_cp" if "dp_cp" in _mesh_resources.root_to_flatten_mapping[device_mesh] else "data_parallel")
```
Comment (cspades) [inline nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:446]: Can we change this to:
```
                    ("dp_cp" if "dp_cp" in _mesh_resources.root_to_flatten_mapping.get(device_mesh, {}) else "data_parallel")
```

This should work as expected when CP=1:
```
(Pdb) _mesh_resources.root_to_flatten_mapping
{}
(Pdb) _mesh_resources.root_to_flatten_mapping.get(device_mesh, {})
{}
(Pdb) device_mesh
DeviceMesh('cuda', [[[0]], [[1]], [[2]], [[3]], [[4]], [[5]], [[6]], [[7]]], mesh_dim_names=('data_parallel', 'context_parallel', 'tensor_parallel'))
```
so we can default to data_parallel if the flattened mesh does not exist.

**Review by akoumpa — APPROVED** [review]
